### PR TITLE
[Enhancement] Support periodic refresh external catalog table

### PR DESCRIPTION
**Pull request by Youngwb**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Support periodic refresh external catalog table, only support HiveTable and HudiTable for now

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 45 (20.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/hudi/HudiMetadata.java|0|1|00.00%|[88]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetadata.java|0|1|00.00%|[83]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetastoreOperations.java|0|1|00.00%|[51]|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|[81]|
|:large_blue_circle:|com/starrocks/connector/hive/CachingHiveMetastore.java|0|3|00.00%|[179, 180, 182]|
|:large_blue_circle:|com/starrocks/server/MetadataMgr.java|1|8|12.50%|[135, 136, 138, 139, 140, 141, 144]|
|:large_blue_circle:|com/starrocks/connector/hive/ConnectorTableMetadataProcessor.java|5|27|18.52%|[63, 64, 67, 68, 69, 70, 73, 74, 75, 76, 77, 78, 79, 80, 82, 83, 84, 85, 86, 87, 88, 89]|
|:large_blue_circle:|com/starrocks/connector/ConnectorMgr.java|3|3|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/connector/ConnectorMetadata.java [hunk 1]
```diff
@@ -69,6 +72,15 @@ default List<String> listTableNames(String dbName) {
         return Lists.newArrayList();
     }
 
+    /**
+     * List all cached table names by caching hive metastore
+     *
+     * @return a set of string containing all cached table names.
+     */
+    default Set<HiveTableName> listCachedTableNames() {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/ConnectorMetadata.java:80]: I think it's better to use CacheUpdateProcessor. this interface isn't common enough

Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/connector/ConnectorMetadata.java:80]: you could refer to the MetastoreEventsProcessor.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 19 / 44 (43.18%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/hive/CacheUpdateProcessor.java|0|1|00.00%|[89]|
|:large_blue_circle:|com/starrocks/connector/hive/CachingHiveMetastore.java|0|1|00.00%|[179]|
|:large_blue_circle:|com/starrocks/connector/hive/ConnectorTableMetadataProcessor.java|12|34|35.29%|[77, 78, 79, 80, 82, 83, 86, 87, 88, 91, 92, 93, 94, 95, 96, 97, 98, 100, 101, 102, 103, 104]|
|:large_blue_circle:|com/starrocks/connector/hudi/HudiConnector.java|3|4|75.00%|[91]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveConnector.java|4|4|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/connector/hive/CachingHiveMetastoreConf.java [hunk 2]
```diff
@@ -34,7 +34,7 @@ public CachingHiveMetastoreConf(Map<String, String> conf) {
         this.cacheRefreshIntervalSec = Long.parseLong(conf.getOrDefault("metastore_cache_refresh_interval_sec",
                 String.valueOf(Config.hive_meta_cache_refresh_interval_s)));
         this.enableListNamesCache = Boolean.parseBoolean(conf.getOrDefault("enable_cache_list_names",
-                "false"));
+                "true"));
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/hive/CachingHiveMetastoreConf.java:37]: it's better disable this config by default.

File: fe/fe-core/src/main/java/com/starrocks/connector/hive/ConnectorTableMetadataProcessor.java [hunk 3]
```diff
@@ -52,6 +67,41 @@ protected void runAfterCatalogReady() {
         }
 
         refreshRegisteredTable();
+        refreshCatalogTable();
+    }
+
+    private void refreshCatalogTable() {
+        MetadataMgr metadataMgr = GlobalStateMgr.getCurrentState().getMetadataMgr();
+        List<String> catalogNames = com.google.common.collect.Lists.newArrayList(cacheUpdateProcessors.keySet());
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/hive/ConnectorTableMetadataProcessor.java:75]: nit: import this class

File: fe/fe-core/src/main/java/com/starrocks/connector/hive/ConnectorTableMetadataProcessor.java [hunk 4]
```diff
@@ -52,6 +67,41 @@ protected void runAfterCatalogReady() {
...
+        for (String catalogName : catalogNames) {
+            CacheUpdateProcessor updateProcessor = cacheUpdateProcessors.get(catalogName);
+            if (updateProcessor == null) {
+                LOG.error("Failed to get cacheUpdateProcessor by catalog {}.", catalogName);
+                continue;
+            }
+            if (CatalogMgr.ResourceMappingCatalog.isResourceMappingCatalog(catalogName)) {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/hive/ConnectorTableMetadataProcessor.java:82]: I think it's better move this check to HiveConnector.onCreate().

File: fe/fe-core/src/main/java/com/starrocks/connector/hive/ConnectorTableMetadataProcessor.java [hunk 5]
```diff
@@ -21,26 +21,41 @@
...
         registeredTableInfos.add(tableInfo);
     }
 
+    public void registerCacheUpdateProcessor(String catalogName, CacheUpdateProcessor cache) {
+        LOG.info("register updating {} metadata cache to ConnectorTableMetadataProcessor", catalogName);
+        cacheUpdateProcessors.put(catalogName, cache);
+    }
+
+    public void unRegisterCacheUpdateProcessor(String catalogName) {
+        LOG.info("stop to register updating {} metadata cache to ConnectorTableMetadataProcessor", catalogName);
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/hive/ConnectorTableMetadataProcessor.java:55]: this log is weird

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 18 / 47 (38.30%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/hive/CacheUpdateProcessor.java|0|4|00.00%|[89, 90, 91, 94]|
|:large_blue_circle:|com/starrocks/connector/hive/CachingHiveMetastore.java|0|1|00.00%|[179]|
|:large_blue_circle:|com/starrocks/connector/hudi/HudiConnector.java|1|5|20.00%|[81, 82, 83, 94]|
|:large_blue_circle:|com/starrocks/connector/hive/ConnectorTableMetadataProcessor.java|12|32|37.50%|[76, 77, 78, 79, 82, 83, 84, 87, 88, 89, 90, 91, 92, 93, 94, 96, 97, 98, 99, 100]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveConnector.java|5|5|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/connector/hive/ConnectorTableMetadataProcessor.java [hunk 6]
```diff
@@ -52,6 +66,38 @@ protected void runAfterCatalogReady() {
...
+            }
+
+            for (HiveTableName cachedTableName : updateProcessor.getCachedTableNames()) {
+                String dbName = cachedTableName.getDatabaseName();
+                String tableName = cachedTableName.getTableName();
+                Table table;
+                try {
+                    table = metadataMgr.getTable(catalogName, dbName, tableName);
+                } catch (Exception e) {
+                    LOG.warn("can't get table of {}.{}.{}", catalogName, dbName, tableName);
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/hive/ConnectorTableMetadataProcessor.java:89]:
```suggestion
                    LOG.warn("can't get table of {}.{}.{}. msg: {}", catalogName, dbName, tableName, e);
```

File: fe/fe-core/src/main/java/com/starrocks/connector/hive/ConnectorTableMetadataProcessor.java [hunk 7]
```diff
@@ -52,6 +66,38 @@ protected void runAfterCatalogReady() {
...
+                    continue;
+                }
+                if (table == null) {
+                    LOG.warn("{}.{}.{} not exist", catalogName, dbName, tableName);
+                    continue;
+                }
+                updateProcessor.refreshTableMetaStoreInfo(dbName, table, true);
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/hive/ConnectorTableMetadataProcessor.java:96]: I think this part should be wrapped in catch. If there are some exceptions, we could get error message from fe.log.

File: fe/fe-core/src/main/java/com/starrocks/connector/hive/HiveConnector.java [hunk 8]
```diff
@@ -89,18 +90,23 @@ private HiveMetadataFactory createMetadataFactory() {
...
+        Optional<CacheUpdateProcessor> updateProcessor = metadataFactory.getCacheUpdateProcessor();
         if (internalMgr.enableHmsEventsIncrementalSync()) {
-            Optional<CacheUpdateProcessor> updateProcessor = metadataFactory.getCacheUpdateProcessor();
             updateProcessor.ifPresent(processor -> GlobalStateMgr.getCurrentState().getMetastoreEventsProcessor()
                     .registerCacheUpdateProcessor(catalogName, updateProcessor.get()));
         }
+        if (!CatalogMgr.ResourceMappingCatalog.isResourceMappingCatalog(catalogName)) {
+            updateProcessor.ifPresent(processor -> GlobalStateMgr.getCurrentState().getConnectorTableMetadataProcessor()
+                    .registerCacheUpdateProcessor(catalogName, updateProcessor.get()));
+        }
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/hive/HiveConnector.java:101]: we could only choose one of the two way of cache.

**Review by stephen-shelby — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 19 / 52 (36.54%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/hive/CacheUpdateProcessor.java|0|4|00.00%|[89, 90, 91, 94]|
|:large_blue_circle:|com/starrocks/connector/hive/CachingHiveMetastore.java|0|1|00.00%|[179]|
|:large_blue_circle:|com/starrocks/connector/hudi/HudiConnector.java|1|5|20.00%|[81, 82, 83, 94]|
|:large_blue_circle:|com/starrocks/connector/hive/ConnectorTableMetadataProcessor.java|12|36|33.33%|[76, 77, 78, 79, 82, 83, 84, 87, 88, 89, 90, 91, 92, 93, 94, 97, 98, 99, 101, 102, 103, 104, 105, 106]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveConnector.java|6|6|100.00%|[]|

**Review by gengjun-git — APPROVED** [review]

**Comment by Youngwb** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**